### ISSUE TITLE
`display_name_method_for` works not as intended with Draper

### ISSUE THREAD

**Issue by faucct** [activeadmin/activeadmin#4279]
This is the breadcrumb I get while using Draper decorator with `delegate_all` option:
![2016-01-21 14 01 40](https://cloud.githubusercontent.com/assets/5202503/12480065/a5a0bf52-c050-11e5-8422-b56779c8fbd9.png)
ActiveAdmin can not look up the display method name correctly here:

``` ruby
      def display_name_method_for(resource)
        @@display_name_methods_cache ||= {}
        @@display_name_methods_cache[resource.class] ||= begin
          methods = active_admin_application.display_name_methods - association_methods_for(resource)
          method  = methods.detect{ |method| resource.respond_to? method }

          if method != :to_s || resource.method(method).source_location
            method
          else
            DISPLAY_NAME_FALLBACK
          end
        end
      end
```

My decorator instance does not respond to any other than `to_s` methods, as the model does. In this case it is supposed to use the fallback, but the problem is that decorator `to_s` method got a source location: `["~/.rvm/gems/ruby-2.1.5/gems/draper-2.1.0/lib/draper/delegation.rb", 10]`.

**Comment by Fivell** [comment]
@faucct , what is your suggestion ? ActiveAdmin provides ```display_name_methods``` option for you and you are able configure it in your app and you're able to define any method from this list in your decorator.

**Comment by faucct** [comment]
I would suggest removing `:to_s` from `display_name_methods` as it is already present in `DISPLAY_NAME_FALLBACK`.

**Comment by Fivell** [comment]
@faucct , look like yes, ```:to_s```  can be removed from ```display_name_methods``` it can be removed, we have specs that it uses to_s after all https://github.com/activeadmin/activeadmin/commit/40e42b3be5be26c0f194bad58b235ec326fdf31a

**Comment by timoschilling** [comment]
Why didn't you simply solve your problem by overriding `to_s` on your decorator or using `display_name`. You can use `display_name` as alias to `to_s` on your `ApplicationRecord`.

**Comment by faucct** [comment]
Are you okay with everyone using decorators solving the problem on their own by copying the code from `DISPLAY_NAME_FALLBACK`?

**Comment by timoschilling** [comment]
Are you ok with breaking existing behavior?

There are many simple ways to solve your problem without copying code or breaking AA.

**Comment by faucct** [comment]
> Are you ok with breaking existing behavior?

Any change is about breaking existing behaviour. Right now I am breaking existing behaviour of `DisplayHelper` rendering `Object#to_s`, e.g. `"#<Post:0x0123456789>"`, for decorators. The "existing behaviour" you are talking about is not covered with tests, so yes, I am ok with breaking this.

> There are many simple ways to solve your problem without copying code or breaking AA.

Are you talking about using the `DISPLAY_NAME_FALLBACK` in my code, because this is the way I have dealt with that issue once popped but I still don't think this is okay to let the AA user deal with this. If this is okay, then you should probably add this solution to the [Using Decorators](https://github.com/activeadmin/activeadmin/blob/master/docs/11-decorators.md) docs.

**Comment by timoschilling** [comment]
I'm talking about define a method on your model or decorator. Your decorator breaks the from AR expected behavior by overwriting AR#to_s

**Comment by faucct** [comment]
I am not sure if I have understood you here. Are you suggesting me defining a method on a model/decorator to fix the problem or are you explaining to me why the problem is arising? Probably the first one since the problem arises even with no methods defined in the decorator.

**Comment by timoschilling** [comment]
Both

**Comment by faucct** [comment]
But my decorator does not override anything: it is just an empty decorator with `delegates_all` as you can see in the added tests, these would fail with a decorator you have in [Decorators](https://github.com/activeadmin/activeadmin/blob/master/docs/11-decorators.md) docs.

**Comment by timoschilling** [comment]
Draper defines a `to_s` method in each decorator. https://github.com/drapergem/draper/blob/master/lib/draper/decorator.rb#L206

**Comment by faucct** [comment]
Yes, and I have stated it at the start of this issue. I am pretty sure that this definition is reasonable and won't be changed inside draper. Are you suggesting that this should be fixed this by putting another guard similar to `resource.method(method).source_location`? Is not my solution simpler and more reliant though it removes some undocumented and untested behaviour?

**Comment by timoschilling** [comment]
Just define a `display_name` method on your model or decorator and every thing is fine

**Comment by timoschilling** [comment]
It is a bug in draper, he delegates a method which didn't exist on the decorated object

**Comment by faucct** [comment]
`Object#to_s` exists in every object thanks to `Kernel#to_s`.

**Comment by timoschilling** [comment]
Yes but for that we have the `resource.method(method).source_location` check

**Comment by timoschilling** [comment]
Why you didn't define a `display_name` method?

**Comment by faucct** [comment]
Well, the check is not perfect and this issue is about it. Do you think we should improve the check, for example if the method is a delegate and the delegated to method does not have a source location?
I did define a display_name method calling `DISPLAY_NAME_FALLBACK` as soon the issue has popped, but should not this be fixed inside AA?

**Comment by timoschilling** [comment]
Should we really fix a bug of draper in AA?

But we can improve the method detection in AA of course. But we should do that in a way that don't depend on a decorator implementation.

**Comment by faucct** [comment]
Is it a bug that draper does not override `Object#method` delegating to the decorated object, so we could check if it is a kernel method?
I have suggested a way we could do this for this specific case of `Draper::Decorator`, but I believe that there are people who use different decorator libraries and it won't work for them, so it is not reliable and will have to be improved forever.

**Comment by timoschilling** [comment]
The problem is that we can check that `to_s` is overwritten, but we can't check how has overwritten it. That's way the solution is to define your own `display_name` method. Yes it's not elegant that everyone needs to do that, but AA can't fix 3 party problems, otherwise we will end up with code like this:
```ruby
          if reource.is?(Draper::Decorator)
            DISPLAY_NAME_FALLBACK
          if reource.is?(Foo::Decorator)
            ...
          if reource.is?(Bar::Decorator)
            ...
          elsif method != :to_s || resource.method(method).source_location
            method
          else
            DISPLAY_NAME_FALLBACK
          end
```

**Comment by faucct** [comment]
We can fix the issue by accepting my PR. The only people affected will be ones relying on `to_s` methods defined in their `ActiveRecord` models to be displayed in activeadmin. Instead it will display `"#{name} ##{id}"` which is not as scary as `"#<Post:0x0123456789>"` is. All they will have to do to bring this back is rename the method to `display_name` or give it the alias if it is used anywhere else.

**Comment by timoschilling** [comment]
Many people have models like this:
```
class User
  def name
    "#{firstname} #{lastname}"
  end

  def to_s
    name
  end
end
```

For that we have the `to_s` in `display_name_methods`, not for `#<Post:0x0123456789>`. I assert that more people overwrite `to_s` then using a decorator.

**Comment by timoschilling** [comment]
What you want is solving your problem by moving the same problem to other people.

**Comment by faucct** [comment]
The behaviour of `display_name` with the `User` model does not change as `name` is still in `display_name_methods`.
Anyway, I am tired of proving my point. Are you okay with this change?

```ruby
def display_name_method_for(resource)
  @@display_name_methods_cache ||= {}
  @@display_name_methods_cache[resource.class] ||= begin
    methods = active_admin_application.display_name_methods - association_methods_for(resource)
    method  = methods.detect{ |method| resource.respond_to? method }

    if method != :to_s || resource.method(method).source_location
      method
    end || DISPLAY_NAME_FALLBACK
  end
end
```

Now the `display_name_method_for` does not expect resource to respond to all methods, so it is possible for me to remove `:to_s` from `display_name_methods`. The existing behaviour you are talking about does not change.

**Comment by deivid-rodriguez** [comment]
@timoschilling Is the last proposal acceptable if it comes with a test ensuring that the existing behavior you talked about is not broken?

**Comment by javierjulio** [comment]
This is something I think we should revisit. We've run into this issue by just implementing Draper decorators. Everything seems to work fine with that integration across many models but this is the one issue that comes up. The problem becomes we would now have to implement a custom display name across many decorators when all we want is the default behavior. I don't know how easy it would be to do here.

**Comment by deivid-rodriguez** [comment]
@javierjulio Did you try @faucct patches, specifically the last one?

**Comment by javierjulio** [comment]
Yes but the result isn't the same. For example, with a multi word named model I noticed the case was different after removing `:to_s` and with or without the patch.

To be clear, just only removing `:to_s` from the `config.display_name_methods` list, without any patch, does resolve the issue.

Regarding the patch though, I'm not familiar with the use of the `||` after the `if` statement. I wasn't sure how that would be different result wise from the original implementation.

**Comment by deivid-rodriguez** [comment]
I see...

> Regarding the patch though, I'm not familiar with the use of the || after the if statement. I wasn't sure how that would be different result wise from the original implementation.

I'm not sure either :thinking:

I'll review this in more depth.

**Comment by AmePelliccia** [comment]
Grandes
